pkg/operator/ceph/file/controller.go

```diff
@@ -0,0 +1,277 @@
+	filesystem, migrationNeeded, err := getFilesystemObject(obj)
+	if err != nil {
+		logger.Errorf("failed to get filesystem object: %+v", err)
+		return
+	}
+
+	if migrationNeeded {
+		logger.Infof("ignoring deletion of legacy filesystem %s in namespace %s", filesystem.Name, filesystem.Namespace)
+		return
+	}
+
+	err = DeleteFilesystem(c.context, *filesystem)
+	if err != nil {
+		logger.Errorf("failed to delete file system %s. %+v", filesystem.Name, err)
+	}
+}
+
+func (c *FilesystemController) filesystemOwners(fs *cephv1alpha1.Filesystem) []metav1.OwnerReference {
+
+	// Only set the cluster crd as the owner of the filesystem resources.
+	// If the filesystem crd is deleted, the operator will explicitly remove the filesystem resources.
+	// If the filesystem crd still exists when the cluster crd is deleted, this will make sure the filesystem
+	// resources are cleaned up.
+	return []metav1.OwnerReference{c.ownerRef}
+}
+
+func filesystemChanged(oldFS, newFS cephv1alpha1.FilesystemSpec) bool {
+	if len(oldFS.DataPools) != len(newFS.DataPools) {
+		logger.Infof("number of data pools changed from %d to %d", len(oldFS.DataPools), len(newFS.DataPools))
+		return true
+	}
+	if oldFS.MetadataServer.ActiveCount != newFS.MetadataServer.ActiveCount {
+		logger.Infof("number of mds active changed from %d to %d", oldFS.MetadataServer.ActiveCount, newFS.MetadataServer.ActiveCount)
+		return true
+	}
+	if oldFS.MetadataServer.ActiveStandby != newFS.MetadataServer.ActiveStandby {
+		logger.Infof("mds active standby changed from %t to %t", oldFS.MetadataServer.ActiveStandby, newFS.MetadataServer.ActiveStandby)
+		return true
+	}
+	return false
+}
+
+func getFilesystemObject(obj interface{}) (filesystem *cephv1alpha1.Filesystem, migrationNeeded bool, err error) {
+	var ok bool
+	filesystem, ok = obj.(*cephv1alpha1.Filesystem)
+	if ok {
+		// the filesystem object is of the latest type, simply return it
+		return filesystem.DeepCopy(), false, nil
+	}
+
+	// type assertion to current filesystem type failed, try instead asserting to the legacy filesystem type
+	// then convert it to the current filesystem type
+	filesystemLegacy, ok := obj.(*rookv1alpha1.Filesystem)
+	if ok {
+		return convertLegacyFilesystem(filesystemLegacy.DeepCopy()), true, nil
+	}
+
+	return nil, false, fmt.Errorf("not a known filesystem object: %+v", obj)
+}
+
+func (c *FilesystemController) migrateFilesystemObject(filesystemToMigrate *cephv1alpha1.Filesystem) error {
+	logger.Infof("migrating legacy filesystem %s in namespace %s", filesystemToMigrate.Name, filesystemToMigrate.Namespace)
+
+	_, err := c.context.RookClientset.CephV1alpha1().Filesystems(filesystemToMigrate.Namespace).Get(filesystemToMigrate.Name, metav1.GetOptions{})
+	if err == nil {
+		// filesystem of current type with same name/namespace already exists, don't overwrite it
+		logger.Warningf("filesystem object %s in namespace %s already exists, will not overwrite with migrated legacy filesystem.",
```

Comment:
we don't support migration across namespaces, right?

Reply:
no, migrations across namespaces is NOT supported.